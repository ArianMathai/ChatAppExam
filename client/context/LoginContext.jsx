import React from "react";

export const LoginContext = React.createContext({
    user: undefined,
    username: undefined,
    google_client_id: undefined,
    google_discovery_url: undefined,
    microsoft_endpoint: undefined,
    microsoft_client_id: undefined,
    loadUser: async () => {},
});