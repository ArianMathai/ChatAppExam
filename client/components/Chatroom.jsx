import {Link} from "react-router-dom";
import {useContext} from "react";
import {LoginContext} from "../context/LoginContext";

function Chatroom(){

    const {user} = useContext(LoginContext);



    return (
        <>
            {user.email ? (
                <div className={"message-board"}>
                    <h2>Chat</h2>
                    <div>
                        <Link to={"/chat/new"}>
                            <button>New chat</button>
                        </Link>
                    </div>
                </div>
            ) : (
                <h2>Login to see chatroom</h2>
            )}
        </>
    );

}
export default Chatroom;