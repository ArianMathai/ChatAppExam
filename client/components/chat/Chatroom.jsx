import {Link, useParams} from "react-router-dom";
import {useContext, useEffect, useRef, useState} from "react";
import {LoginContext} from "../../context/LoginContext";
import MessageInput from "./MessageInput";
import {WebSocketContext} from "../webSocket/WebSocketProvider";

function Chatroom(){

    const {user} = useContext(LoginContext);
    const {username} = useContext(LoginContext);
    const [messages, setMessages] = useState([])
    const [room, setRoom] = useState();
    const webSocket = useContext(WebSocketContext);
    const messageRef = useRef(null);

    const { roomName } = useParams();
    console.log("Roomname in chatroom = ", roomName, room)

    async function fetchRoom(){
        const res = await fetch(`/api/chat/getRoom/${roomName}`, )
        const data = await res.json();
        console.log("Data = room = ", data)

        if (res.ok && data) {
            console.log("setting room")
            console.log(data)
            setRoom(data);
            fetchMessages();
        } else {
            console.log("Did not find room");
        }

    }

    useEffect(() => {
        fetchRoom()
    }, []);

    useEffect(() => {
        console.log("in useEffect with socket")

        const handleWebSocketMessage = (event) => {
            const data = JSON.parse(event.data);

            console.log("Received message from server:", data);
            setMessages((prevMessages) => [...prevMessages, data.message]);
        };

        if (webSocket) {
            webSocket.addEventListener("message", handleWebSocketMessage);

        }
    }, [webSocket]);

    useEffect(() => {
        fetchMessages();
    }, []);

    useEffect(() => {
        if (messageRef.current) {
            messageRef.current.scrollTop = messageRef.current.scrollHeight;
        }
    }, [messages]);



    async function fetchMessages(){

        const res = await fetch(`/api/chat/getAllMessages/${roomName}`, {

        })
        const msg = await res.json();
        console.log("MSG = ", msg.messages)

        if (res.ok){
            setMessages(msg.messages);
        } else {
            console.log("Did not retrieve messages")
        }

    }



    return (
        <>
            {username ? (
                <div className={"message-board"}>
                    {room ? (
                        <>
                            <div className="message-board">
                                <h2>Chatroom: {room.room.roomName}</h2>
                                <div className="messageList" ref={messageRef}>
                                    {messages.map((message, index) => (
                                        <div className="message-container" key={index}>
                                            <p className="message-line">{message.username} : {message.message}</p>
                                        </div>
                                    ))}
                                </div>
                                <div className="message-input">
                                    <label htmlFor="messageInput"></label>
                                    <MessageInput room={room} />
                                </div>
                                <div className="button-container">
                                    <Link to={"/chat/new"}>
                                        <button>New chat</button>
                                    </Link>
                                </div>
                            </div>


                        </>
                    ) : (
                        <p>Loading room...</p>
                    )}

                </div>
            ) : (
                <h2>Login to see chatroom</h2>
            )}
        </>
    );


}
export default Chatroom;