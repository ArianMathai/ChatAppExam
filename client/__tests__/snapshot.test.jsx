import renderer from "react-test-renderer";
import {LoginContext} from "../context/LoginContext";
import Chatroom from "../components/chat/Chatroom";
import {MemoryRouter, Route, Routes} from "react-router-dom";





describe("frontend application react", () => {
    it("should render list of messages", async () => {
        const username = "brukernavn";
        const messages = [
            { username: "loyd", message: "testing testing!" },
            { username: "frank", message: "Er det noen her?" },
        ];


        jest.spyOn(global, "fetch").mockImplementation(() =>
            Promise.resolve({
                json: () => Promise.resolve({ messages }),
                ok: true,
            })
        );

        let component;

        await act(async () => {
            component = renderer.create(
                <MemoryRouter initialEntries={["/chatroom/test"]}>
                    <Routes>
                        <Route
                            path="/chatroom/:roomName"
                            element={
                                <LoginContext.Provider value={{ username }}>
                                    <Chatroom />
                                </LoginContext.Provider>
                            }
                        />
                    </Routes>
                </MemoryRouter>
            );
        });

        expect(component).toMatchSnapshot();
    });

    it("renders correctly if not logged in", () => {
        const component = renderer.create(
            <MemoryRouter>
                <Chatroom />
            </MemoryRouter>
        );

        const t = component.toJSON();
        expect(t).toMatchSnapshot();
    });

    it("renders correctly when loading room", () => {
        const username = "brukernavn";
        const component = renderer.create(
            <MemoryRouter>
                <LoginContext.Provider value={{ username }}>
                    <Chatroom />
                </LoginContext.Provider>
            </MemoryRouter>
        );

        const x = component.toJSON();
        expect(x).toMatchSnapshot();
    });


});
