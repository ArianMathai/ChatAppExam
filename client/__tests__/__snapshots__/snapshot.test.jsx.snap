// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`frontend application react renders correctly 1`] = `
<h2>
  Login to see chatroom
</h2>
`;

exports[`frontend application react renders correctly for room 'exampleRoom' 1`] = `
<div
  className="message-board"
>
  <p>
    Loading room...
  </p>
</div>
`;


